docs/rfc/0000-cli-infrastructure.adoc

```diff
@@ -0,0 +1,237 @@
+= RFC: RadiCLI
+:author: @FintanH
+:revdate: 2021-06-04
+:revremark: draft
+:toc:
+:toc-placement: preamble
+
+* Author: {author}
+* Date: {revdate}
+* Status: {revremark}
+* Community discussion: n/a
+* Tracking Issue: n/a
+
+== Motivation
+
+RFC-682 set the stage for defining the architecture of applications
+built on top of `<<radicle-link>>`. This RFC wishes to build on top of
+this specification and fill in the gaps around the CLI component and
+its organisation. As dogfooding our own solutions is on our long-term
+roadmap, this team wants to provide tooling that they will maintain
+and use themselves.
+
+== Overview
+
+This proposal will outline an binary called `rad`, what it will
+provide, and how it can be extended. Following this we provide details
+on the interaction -- or lack there of -- with the peer-to-peer
+daemon.
+
+We also provide a proposal of how any CLI tool will have
+access to any keys, as well as what needs to be completed for
+providing this.
+
+We close out with how the core components will be organised within
+`radicle-link` and `<<radicle-bins>>` as well as providing the first
+iteration of the core set of subcommands.
+
+== One `rad` to Rule Them All
+
+[quote]
+-------
+One ring to rule them all, one ring to find them,
+One ring to bring them all, and in the darkness bind them;
+-------
+-- J.R.R. Tolkein, Fellowship of the Ring
+
+The integral part of this proposal is that there will be one binary,
+`rad`, that will be the foundation for all other subcommands. It will
+define the global parameters and how to extend it.
+
+We will discuss these in the following two sections.
+
+== The Fellowship of the `rad`
+
+[quoute]
+--------
+“I can’t carry it for you, but I can carry you.” – Sam Gamgee
+--------
+-- J.R.R. Tolkien, The Return of the King
+
+To allow the ecosystem of commands to be extended by ourselves and
+others in the community it will be necessary to extend the `rad`
+command with further subcommands. The details of this are outlined in
+RFC-682#CLI, however, we will give a quick overview here for the sake
+of the reader.
+
+Taking inspiration from tools such as `git` and `cargo`, we propose
+that the `rad` binary can be extended by placing another
+binary on the users `$PATH` (or `%PATH%` on Windows). This
+binary MUST follow the naming convention of `rad-*`. To reiterate
+here, these new commands MUST NOT override the core set outlined here
+(nor in further RFCs).
+
+== Global Parameters
+
+Many subcommands will likely want to share a set of options. These
+options, dubbed as "global parameters" here, will be a defined set --
+exported from `radicle-link` for re-use -- that the base `rad` command
+will always receive and pass onto the subcommand being called.
+
+Thus, all subcommands SHALL receive these global parameters as part of
+their interface. We list these parameters here, but it is NOT an
+exhaustive list:
+
+* `verbose`
+* `quiet`
+* `profile := active (default) | <profile-id>`
```

Comment:
How does this relate to location of the monorepo (i.e. `RAD_HOME`)? Surfacing the relationship would be interesting in the context of this RFC.

Reply:
Ya, I was thinking about this. I think there's no need to surface `RAD_HOME` since:
a) it can be set on the command-line, i.e. `RAD_HOME=/home/happy/foo rad key create`
b) the general case should be that `XDG_DATA_HOME` is used, and `RAD_HOME` is used in special cases. Contrasting that with the profiles, which can also be set via `RAD_PROFILE`, these will potentially be created and set more frequently